File: app/src/main/java/com/qwe7002/telegram_rc/beacon_receiver_service.java
package com.qwe7002.telegram_rc;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.net.wifi.WifiManager;
import android.os.BatteryManager;
import android.os.IBinder;
import android.os.PowerManager;
import android.os.RemoteException;
import android.provider.Settings;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.fitc.wifihotspot.TetherManager;
import com.google.gson.Gson;
import com.qwe7002.telegram_rc.config.beacon;
import com.qwe7002.telegram_rc.data_structure.beacon_list;
import com.qwe7002.telegram_rc.data_structure.request_message;
import com.qwe7002.telegram_rc.static_class.CONST;
import com.qwe7002.telegram_rc.static_class.network;
import com.qwe7002.telegram_rc.static_class.notify;
import com.qwe7002.telegram_rc.static_class.other;
import com.qwe7002.telegram_rc.static_class.remote_control;
import com.qwe7002.telegram_rc.static_class.resend;

import org.altbeacon.beacon.Beacon;
import org.altbeacon.beacon.BeaconConsumer;
import org.altbeacon.beacon.BeaconManager;
import org.altbeacon.beacon.BeaconParser;
import org.altbeacon.beacon.Region;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Objects;

import io.paperdb.Paper;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class beacon_receiver_service extends Service {
    private final static String TAG = "beacon_receviver";
    private Context context;
    private WifiManager wifi_manager;
    private int not_found_count = 0;
    private int detect_singal_count = 0;
    private OkHttpClient okhttp_client;
    private String chat_id;
    private String request_url;
    static String message_thread_id;
    private BeaconManager beacon_manager;
    private long startup_time = 0;
    private long last_receive_time = 0;
    private long last_action_time = 0;
    private beacon config;
    private beacon_service_consumer beacon_consumer;
    private PowerManager.WakeLock wakelock;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @NotNull
    private batteryInfo getBatteryInfo() {
        batteryInfo info = new batteryInfo();
        BatteryManager batteryManager = (BatteryManager) context.getSystemService(BATTERY_SERVICE);
        assert batteryManager != null;
        info.battery_level = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY);
        IntentFilter intentfilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
        Intent batteryStatus = context.registerReceiver(null, intentfilter);
        assert batteryStatus != null;
        int charge_status = batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1);

        switch (charge_status) {
            case BatteryManager.BATTERY_STATUS_CHARGING:
            case BatteryManager.BATTERY_STATUS_FULL:
                info.is_charging = true;
                break;
            case BatteryManager.BATTERY_STATUS_DISCHARGING:
            case BatteryManager.BATTERY_STATUS_NOT_CHARGING:
                switch (batteryStatus.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1)) {
                    case BatteryManager.BATTERY_PLUGGED_AC:
                    case BatteryManager.BATTERY_PLUGGED_USB:
                    case BatteryManager.BATTERY_PLUGGED_WIRELESS:
                        info.is_charging = true;
                        break;
                }
                break;
        }
        return info;
    }

    static class batteryInfo {
        int battery_level = 0;
        boolean is_charging = false;
    }


    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private final BroadcastReceiver reloadConfigReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            config = Paper.book("beacon_config").read("config", new beacon());
            beacon_manager.setForegroundScanPeriod(config.delay);
            beacon_manager.setForegroundBetweenScanPeriod(2000);
            try {
                beacon_manager.updateScanPeriods();
            } catch (RemoteException e) {
                e.printStackTrace();
            }
            startup_time = System.currentTimeMillis();
        }
    };

    @SuppressLint({"InvalidWakeLockTag", "WakelockTimeout"})
    @Override
    public void onCreate() {
        super.onCreate();
        context = getApplicationContext();
        Paper.init(context);

        wakelock = ((PowerManager) Objects.requireNonNull(context.getSystemService(Context.POWER_SERVICE))).newWakeLock(android.os.PowerManager.PARTIAL_WAKE_LOCK, "beacon_receive");
        wakelock.setReferenceCounted(false);
        if (!wakelock.isHeld()) {
            wakelock.acquire();
        }

        config = Paper.book("beacon_config").read("config", new beacon());
        LocalBroadcastManager.getInstance(this).registerReceiver(reloadConfigReceiver,
                new IntentFilter("reload_beacon_config"));
        SharedPreferences sharedPreferences = context.getSharedPreferences("data", MODE_PRIVATE);
        request_url = network.get_url(sharedPreferences.getString("bot_token", ""), "SendMessage");
        chat_id = sharedPreferences.getString("chat_id", "");
        message_thread_id = sharedPreferences.getString("message_thread_id", "");
        wifi_manager = (WifiManager) getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        okhttp_client = network.get_okhttp_obj(sharedPreferences.getBoolean("doh_switch", true));
        beacon_consumer = new beacon_service_consumer();
        beacon_manager = BeaconManager.getInstanceForApplication(this);
        // Detect iBeacon:
        beacon_manager.getBeaconParsers().add(new BeaconParser().
                setBeaconLayout("m:2-3=0215,i:4-19,i:20-21,i:22-23,p:24-24"));

        // Detect the main identifier (UID) frame:
        beacon_manager.getBeaconParsers().add(new BeaconParser().
                setBeaconLayout(BeaconParser.EDDYSTONE_UID_LAYOUT));

        // Detect the telemetry (TLM) frame:
        beacon_manager.getBeaconParsers().add(new BeaconParser().
                setBeaconLayout(BeaconParser.EDDYSTONE_TLM_LAYOUT));

        // Detect the URL frame:
        beacon_manager.getBeaconParsers().add(new BeaconParser().
                setBeaconLayout(BeaconParser.EDDYSTONE_URL_LAYOUT));
        Notification notification = other.get_notification_obj(context, getString(R.string.beacon_receiver));
        beacon_manager.setEnableScheduledScanJobs(false);
        beacon_manager.enableForegroundServiceScanning(notification, notify.BEACON_SERVICE);
        startForeground(notify.BEACON_SERVICE, notification);
        //noinspection deprecation
        BeaconManager.setAndroidLScanningDisabled(false);
        beacon_manager.setForegroundScanPeriod(config.delay);
        beacon_manager.setForegroundBetweenScanPeriod(2000);
        startup_time = System.currentTimeMillis();
        //noinspection deprecation
        beacon_manager.bind(beacon_consumer);
    }

    @Override
    public void onDestroy() {
        //noinspection deprecation
        beacon_manager.unbind(beacon_consumer);
        beacon_manager.disableForegroundServiceScanning();
        wakelock.release();
        stopForeground(true);
        super.onDestroy();
    }

    private void network_progress_handle(String message, String chat_id, @NotNull OkHttpClient okhttp_client) {
        request_message request_body = new request_message();
        request_body.chat_id = chat_id;
        request_body.message_thread_id = message_thread_id;
        request_body.text = message + "\n" + context.getString(R.string.current_battery_level) + chat_command_service.get_battery_info(context) + "\n" + getString(R.string.current_network_connection_status) + chat_command_service.get_network_type(context);
        String request_body_json = new Gson().toJson(request_body);
        RequestBody body = RequestBody.create(request_body_json, CONST.JSON);
        Request request_obj = new Request.Builder().url(request_url).method("POST", body).build();
        Call call = okhttp_client.newCall(request_obj);
        call.enqueue(new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                resend.add_resend_loop(context, request_body.text);
                e.printStackTrace();
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                Log.d(TAG, "onResponse: " + Objects.requireNonNull(response.body()).string());
            }
        });
    }

    @SuppressWarnings("deprecation")
    class beacon_service_consumer implements BeaconConsumer {
        @Override
        public void onBeaconServiceConnect() {
            beacon_manager.addRangeNotifier((beacons, region) -> {
                beacon_list.beacons = beacons;
                boolean wifi_is_enable_status;
                LocalBroadcastManager.getInstance(context).sendBroadcast(new Intent("flush_view"));
                if (config.use_vpn_hotspot) {
                    if (!remote_control.is_vpn_hotsport_exist(context) && Settings.System.canWrite(context)) {
                        config.use_vpn_hotspot = false;
                    }
                    wifi_is_enable_status = Paper.book("temp").read("wifi_open", false);
                } else {
                    if (!Settings.System.canWrite(context) && remote_control.is_vpn_hotsport_exist(context)) {
                        config.use_vpn_hotspot = true;
                    }
                    wifi_is_enable_status = remote_control.is_tether_active(context);
                }
                long current_time = System.currentTimeMillis();
                if ((System.currentTimeMillis() - startup_time) < 10000L) {
                    Log.d(TAG, "onBeaconServiceConnect: Startup time is too short");
                    not_found_count = 0;
                    detect_singal_count = 0;
                    return;
                }
                if ((current_time - last_receive_time) < config.delay) {
                    Log.d(TAG, "onBeaconServiceConnect: Last receive boardcast time is too short");
                    return;
                }
                last_receive_time = current_time;
                if ((current_time - last_action_time) <= 5000L) {
                    not_found_count = 0;
                    detect_singal_count = 0;
                    Log.d(TAG, "onBeaconServiceConnect: Last action time is too short");
                    return;
                }
                if (Paper.book().read("disable_beacon", false)) {
                    not_found_count = 0;
                    detect_singal_count = 0;
                    return;
                }
                batteryInfo info = getBatteryInfo();
                if (!info.is_charging && info.battery_level < 25 && !wifi_is_enable_status && !config.opposite) {
                    not_found_count = 0;
                    detect_singal_count = 0;
                    Log.d(TAG, "onBeaconServiceConnect: Turn off beacon automatic activation");

                    return;
                }
                ArrayList<String> listen_beacon_list = Paper.book("beacon_config").read("address", new ArrayList<>());
                if (listen_beacon_list.size() == 0) {
                    not_found_count = 0;
                    detect_singal_count = 0;
                    Log.i(TAG, "onBeaconServiceConnect: Watchlist is empty");
                    return;
                }
                boolean found_beacon = false;
                Beacon detect_beacon = null;
                for (Beacon beacon : beacons) {
                    Log.d(TAG, "Mac address: " + beacon.getBluetoothAddress() + " Rssi: " + beacon.getRssi() + " Power: " + beacon.getTxPower() + " Distance: " + beacon.getDistance());
                    for (String beacon_address : listen_beacon_list) {
                        if (beacon.getBluetoothAddress().equals(beacon_address)) {
                            found_beacon = true;
                            break;
                        }
                    }
                    if (found_beacon) {
                        detect_beacon = beacon;
                        break;
                    }
                }

                String beacon_status;
                if (found_beacon) {
                    beacon_status = "\nBeacon Rssi: " + detect_beacon.getRssi() + "dBm";
                    not_found_count = 0;
                    if (detect_beacon.getRssi() < config.RSSI_strenght) {
                        Log.i(TAG, "onBeaconServiceConnect: Signal is too weak, no operation");
                    } else {
                        ++detect_singal_count;
                    }
                } else {
                    beacon_status = "\nBeacon Not Found.";
                    if (not_found_count > 1) {
                        detect_singal_count = 0;
                    }

                    ++not_found_count;
                }

                final int STATUS_STANDBY = -1;
                final int STATUS_ENABLE_AP = 0;
                final int STATUS_DISABLE_AP = 1;
                int switch_status = STATUS_STANDBY;
                Log.d(TAG, "detect_singal_count: " + detect_singal_count);
                Log.d(TAG, "not_found_count: " + not_found_count);
                if (wifi_is_enable_status && found_beacon) {
                    if (!config.opposite) {
                        if (detect_singal_count >= config.disable_count) {
                            last_action_time = current_time;
                            detect_singal_count = 0;
                            not_found_count = 0;
                            if (config.use_vpn_hotspot) {
                                remote_control.disable_vpn_ap(wifi_manager);
                            } else {
                                remote_control.disable_tether(context, TetherManager.TetherMode.TETHERING_WIFI);
                            }
                            switch_status = STATUS_DISABLE_AP;
                        }
                    } else {
                        if (detect_singal_count >= config.enable_count) {
                            last_action_time = current_time;
                            detect_singal_count = 0;
                            not_found_count = 0;
                            if (config.use_vpn_hotspot) {
                                remote_control.enable_vpn_ap(wifi_manager);
                            } else {
                                remote_control.enable_tether(context, TetherManager.TetherMode.TETHERING_WIFI);
                            }
                            switch_status = STATUS_ENABLE_AP;
                        }
                    }
                }
                if (!wifi_is_enable_status && !found_beacon) {
                    if (!config.opposite) {
                        if (not_found_count >= config.enable_count) {
                            last_action_time = current_time;
                            detect_singal_count = 0;
                            not_found_count = 0;
                            if (config.use_vpn_hotspot) {
                                remote_control.enable_vpn_ap(wifi_manager);
                            } else {
                                remote_control.enable_tether(context, TetherManager.TetherMode.TETHERING_WIFI);
                            }
                            switch_status = STATUS_ENABLE_AP;
                        }
                    } else {
                        if (not_found_count >= config.disable_count) {
                            last_action_time = current_time;
                            detect_singal_count = 0;
                            not_found_count = 0;
                            if (config.use_vpn_hotspot) {
                                remote_control.disable_vpn_ap(wifi_manager);
                            } else {
                                remote_control.disable_tether(context, TetherManager.TetherMode.TETHERING_WIFI);
                            }
                            switch_status = STATUS_DISABLE_AP;
                        }
                    }
                }
                String message = null;
                switch (switch_status) {
                    case STATUS_ENABLE_AP:
                        message = getString(R.string.system_message_head) + "\n" + getString(R.string.enable_wifi) + getString(R.string.action_success) + beacon_status;
                        break;
                    case STATUS_DISABLE_AP:
                        message = getString(R.string.system_message_head) + "\n" + getString(R.string.disable_wifi) + getString(R.string.action_success) + beacon_status;
                        break;
                }
                if (message != null) {
                    network_progress_handle(message, chat_id, okhttp_client);
                }
            });

            beacon_manager.startRangingBeacons(new Region("com.qwe7002.telegram_rc", null, null, null));
        }


        @Override
        public Context getApplicationContext() {
            return beacon_receiver_service.this.getApplicationContext();
        }

        @Override
        public void unbindService(ServiceConnection serviceConnection) {
            beacon_receiver_service.this.unbindService(serviceConnection);
        }

        @Override
        public boolean bindService(Intent intent, ServiceConnection serviceConnection, int i) {
            return beacon_receiver_service.this.bindService(intent, serviceConnection, i);
        }

    }
}

File: app/src/main/java/com/qwe7002/telegram_rc/beacon_receiver_service.kt
@file:Suppress("DEPRECATION")

package com.qwe7002.telegram_rc

import aga.android.luch.BeaconScanner
import aga.android.luch.IBeaconBatchListener
import aga.android.luch.IScanner
import aga.android.luch.ScanDuration
import aga.android.luch.parsers.BeaconParserFactory
import android.Manifest
import android.annotation.SuppressLint
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.net.wifi.WifiManager
import android.os.BatteryManager
import android.os.IBinder
import android.os.PowerManager
import android.os.PowerManager.WakeLock
import android.provider.Settings
import android.telephony.TelephonyManager
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.fitc.wifihotspot.TetherManager
import com.google.gson.Gson
import com.qwe7002.telegram_rc.config.beacon
import com.qwe7002.telegram_rc.data_structure.BeaconModel
import com.qwe7002.telegram_rc.data_structure.beaconList
import com.qwe7002.telegram_rc.data_structure.beaconList.beacons
import com.qwe7002.telegram_rc.data_structure.request_message
import com.qwe7002.telegram_rc.root_kit.radio
import com.qwe7002.telegram_rc.static_class.CONST
import com.qwe7002.telegram_rc.static_class.network
import com.qwe7002.telegram_rc.static_class.notify
import com.qwe7002.telegram_rc.static_class.other
import com.qwe7002.telegram_rc.static_class.remote_control
import com.qwe7002.telegram_rc.static_class.resend
import io.paperdb.Paper
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.util.Objects


class beacon_receiver_service : Service() {
    private val TAG = "beacon_receviver"
    private lateinit var wifiManager: WifiManager
    private var notFoundCount = 0
    private var detectCount = 0
    private lateinit var okhttpClient: OkHttpClient
    private lateinit var chatId: String
    private lateinit var requestUrl: String
    lateinit var messageThreadId: String
    private lateinit var scanner: IScanner
    private lateinit var config: beacon
    private lateinit var wakelock: WakeLock
    override fun onBind(intent: Intent): IBinder {
        TODO("Return the communication channel to the service.")
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return START_STICKY
    }

    private val reloadConfigReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            config = Paper.book("beacon_config").read("config", beacon())!!
        }
    }

    @SuppressLint("InvalidWakeLockTag", "WakelockTimeout")
    override fun onCreate() {
        super.onCreate()
        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            Log.d(TAG, "onCreate: permission denied")
            return
        }
        Paper.init(applicationContext)
        wakelock =
            (Objects.requireNonNull(applicationContext.getSystemService(POWER_SERVICE)) as PowerManager).newWakeLock(
                PowerManager.PARTIAL_WAKE_LOCK, "beacon_receive"
            )
        wakelock.setReferenceCounted(false)
        if (!this.wakelock.isHeld) {
            this.wakelock.acquire()
        }
        config = Paper.book("beacon_config").read("config", beacon())!!
        LocalBroadcastManager.getInstance(this).registerReceiver(
            reloadConfigReceiver,
            IntentFilter("reload_beacon_config")
        )
        val sharedPreferences = applicationContext.getSharedPreferences("data", MODE_PRIVATE)
        requestUrl = network.getUrl(sharedPreferences.getString("bot_token", ""), "SendMessage")
        chatId = sharedPreferences.getString("chat_id", "")!!
        messageThreadId = sharedPreferences.getString("message_thread_id", "")!!
        wifiManager = applicationContext.getSystemService(WIFI_SERVICE) as WifiManager
        okhttpClient = network.getOkhttpObj(sharedPreferences.getBoolean("doh_switch", true))
        val notification =
            other.getNotificationObj(applicationContext, getString(R.string.beacon_receiver))
        startForeground(notify.BEACON_SERVICE, notification)

        val batchListener = IBeaconBatchListener { beacons ->
            beaconList.beacons = ArrayList()
            beacons.toList().map {
                val item = BeaconModel(
                    uuid = it.getIdentifierAsUuid(1).toString(),
                    major = it.getIdentifierAsInt(2),
                    minor = it.getIdentifierAsInt(3),
                    rssi = it.rssi.toInt(),
                    hardwareAddress = it.hardwareAddress,
                    distance = scanner.ranger.calculateDistance(it)
                )
                beaconList.beacons.add(item)
                Log.d(TAG, "onCreate: $item")
            }
            LocalBroadcastManager.getInstance(applicationContext)
                .sendBroadcast(Intent("flush_beacons_list"));
        }
        val beaconLayout =
            "m:2-3=0215,i:4-19,i:20-21,i:22-23,p:24-24" // search the Internet to find the layout string of your specific beacon

        val beaconParser = BeaconParserFactory.createFromLayout(beaconLayout)

        scanner = BeaconScanner.Builder(application)
            .setBeaconParser(beaconParser)
            .setBeaconExpirationDuration(10)
            .setScanDuration(
                ScanDuration.UNIFORM
            )
            .setRangingEnabled()
            .setBeaconBatchListener(batchListener)
            .build()

        LocalBroadcastManager.getInstance(this).registerReceiver(
            flushReceiver,
            IntentFilter("flush_beacons_list")
        )
        scanner.start()

    }

    private val flushReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val wifiIsEnableStatus: Boolean
            if (config.useVpnHotspot) {
                if (!remote_control.isVPNHotspotExist(context) && Settings.System.canWrite(
                        context
                    )
                ) {
                    config.useVpnHotspot = false;
                }
                wifiIsEnableStatus = Paper.book("temp").read("wifi_open", false)!!;
            } else {
                if (!Settings.System.canWrite(context) && remote_control.isVPNHotspotExist(
                        context
                    )
                ) {
                    config.useVpnHotspot = true;
                }
                wifiIsEnableStatus = remote_control.isHotspotActive(context);
            }
            if (Paper.book().read("disable_beacon", false)!!) {
                notFoundCount = 0;
                detectCount = 0;
                return;
            }
            val info = getBatteryInfo()
            if (!info.isCharging && info.batteryLevel < 25 && !wifiIsEnableStatus && !config.opposite) {
                notFoundCount = 0
                detectCount = 0
                Log.d(TAG, "onBeaconServiceConnect: Turn off beacon automatic activation")
                return
            }
            val listenBeaconList =
                Paper.book("beacon_config").read("address", ArrayList<String>())!!
            if (listenBeaconList.size == 0) {
                notFoundCount = 0
                detectCount = 0
                Log.i(TAG, "onBeaconServiceConnect: Watchlist is empty")
                return
            }
            var foundBeacon = false
            var detectBeacon: BeaconModel? = null
            for (beacon in beacons) {
                Log.d(
                    TAG,
                    "UUID: " + beacon.uuid + " Rssi: " + beacon.rssi
                )
                for (beacon_address in listenBeaconList) {

                    if (beaconList.beaconItemName(beacon.uuid, beacon.major, beacon.minor)
                            .equals(beacon_address)
                    ) {
                        foundBeacon = true
                        break
                    }
                }
                if (foundBeacon) {
                    detectBeacon = beacon
                    break
                }
            }
            var beaconStatus = ""
            if (foundBeacon) {
                if (detectBeacon != null) {
                    val distance = detectBeacon.distance.toInt().toString()
                    beaconStatus = "\nBeacon Distance: $distance meter"
                    notFoundCount = 0
                    ++detectCount
                };
            } else {
                beaconStatus = "\nBeacon Not Found."
                if (notFoundCount > 1) {
                    detectCount = 0
                }
                ++notFoundCount
            }
            val STATUS_STANDBY = -1
            val STATUS_ENABLE_AP = 0
            val STATUS_DISABLE_AP = 1
            var switchStatus = STATUS_STANDBY
            Log.d(TAG, "detect_singal_count: $detectCount")
            Log.d(TAG, "not_found_count: $notFoundCount")
            if (wifiIsEnableStatus && foundBeacon) {
                if (!config.opposite) {
                    if (detectCount >= config.disableCount) {
                        detectCount = 0
                        notFoundCount = 0
                        if (config.useVpnHotspot) {
                            remote_control.disableVPNHotspot(wifiManager)
                        } else {
                            remote_control.disableHotspot(
                                context,
                                TetherManager.TetherMode.TETHERING_WIFI
                            )
                        }
                        switchStatus = STATUS_DISABLE_AP
                    }
                } else {
                    if (detectCount >= config.enableCount) {
                        detectCount = 0
                        notFoundCount = 0
                        if (config.useVpnHotspot) {
                            remote_control.enableVPNHotspot(wifiManager)
                        } else {
                            remote_control.enableHotspot(
                                context,
                                TetherManager.TetherMode.TETHERING_WIFI
                            )
                        }
                        switchStatus = STATUS_ENABLE_AP
                    }
                }
            }
            if (!wifiIsEnableStatus && !foundBeacon) {
                if (!config.opposite) {
                    if (notFoundCount >= config.enableCount) {
                        detectCount = 0
                        notFoundCount = 0
                        if (config.useVpnHotspot) {
                            remote_control.enableVPNHotspot(wifiManager)
                        } else {
                            remote_control.enableHotspot(
                                context,
                                TetherManager.TetherMode.TETHERING_WIFI
                            )
                        }
                        switchStatus = STATUS_ENABLE_AP
                    }
                } else {
                    if (notFoundCount >= config.disableCount) {
                        detectCount = 0
                        notFoundCount = 0
                        if (config.useVpnHotspot) {
                            remote_control.disableVPNHotspot(wifiManager)
                        } else {
                            remote_control.disableHotspot(
                                context,
                                TetherManager.TetherMode.TETHERING_WIFI
                            )
                        }
                        switchStatus = STATUS_DISABLE_AP
                    }
                }
            }

            var message: String? = null
            when (switchStatus) {
                STATUS_ENABLE_AP -> message = "${getString(R.string.system_message_head)}\n${getString(R.string.enable_wifi)}${getString(R.string.action_success)}$beaconStatus"

                STATUS_DISABLE_AP -> message = "${getString(R.string.system_message_head)}\n${getString(R.string.disable_wifi)}${getString(R.string.action_success)}$beaconStatus"
            }
            message?.let { networkHandle(it, chatId, okhttpClient) }

        }
    }

    override fun onDestroy() {
        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            scanner.stop()
        }
        wakelock.release()
        this.stopForeground(true)
        super.onDestroy()
    }


    private fun networkHandle(
        message: String,
        chat_id: String,
        okhttp_client: OkHttpClient
    ) {
        val request_body = request_message()
        request_body.chat_id = chat_id
        request_body.message_thread_id = messageThreadId
        request_body.text = message + "\n" + getString(R.string.current_battery_level) + getBatteryInfoMsg() + "\n" + getString(R.string.current_network_connection_status) +networkType();
        val requestBodyJson = Gson().toJson(request_body)
        val body: RequestBody = requestBodyJson.toRequestBody(CONST.JSON)
        val requestObj: Request = Request.Builder().url(requestUrl).method("POST", body).build()
        val call: Call = okhttp_client.newCall(requestObj)
        call.enqueue(object : Callback {

            override fun onFailure(call: Call, e: IOException) {
                resend.addResendLoop(applicationContext, request_body.text)
                e.printStackTrace()
            }

            override fun onResponse(call: Call, response: Response) {
                Log.d(TAG, "onResponse: " + (response.body?.string() ?: ""))
            }
        })
    }

    fun getBatteryInfoMsg(): String {
        val batteryManager = (applicationContext.getSystemService(BATTERY_SERVICE) as BatteryManager)
        var battery_level = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
        if (battery_level > 100) {
            Log.i(
                "get_battery_info",
                "The previous battery is over 100%, and the correction is 100%."
            )
            battery_level = 100
        }
        val intentfilter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        val batteryStatus = applicationContext.registerReceiver(null, intentfilter)!!
        val charge_status = batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
        val battery_string_builder = StringBuilder().append(battery_level).append("%")
        when (charge_status) {
            BatteryManager.BATTERY_STATUS_CHARGING, BatteryManager.BATTERY_STATUS_FULL -> battery_string_builder.append(
                " ("
            ).append(applicationContext.getString(R.string.charging)).append(")")

            BatteryManager.BATTERY_STATUS_DISCHARGING, BatteryManager.BATTERY_STATUS_NOT_CHARGING -> when (batteryStatus.getIntExtra(
                BatteryManager.EXTRA_PLUGGED,
                -1
            )) {
                BatteryManager.BATTERY_PLUGGED_AC, BatteryManager.BATTERY_PLUGGED_USB, BatteryManager.BATTERY_PLUGGED_WIRELESS -> battery_string_builder.append(
                    " ("
                ).append(applicationContext.getString(R.string.not_charging)).append(")")
            }
        }
        return battery_string_builder.toString()
    }
    private fun getBatteryInfo(): BatteryInfo {
        val info = BatteryInfo()
        val batteryManager =
            (applicationContext.getSystemService(BATTERY_SERVICE) as BatteryManager)
        info.batteryLevel = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
        val intent = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        val batteryStatus: Intent = applicationContext.registerReceiver(null, intent)!!
        when (batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1)) {
            BatteryManager.BATTERY_STATUS_CHARGING, BatteryManager.BATTERY_STATUS_FULL -> info.isCharging =
                true

            BatteryManager.BATTERY_STATUS_DISCHARGING, BatteryManager.BATTERY_STATUS_NOT_CHARGING -> when (batteryStatus.getIntExtra(
                BatteryManager.EXTRA_PLUGGED,
                -1
            )) {
                BatteryManager.BATTERY_PLUGGED_AC, BatteryManager.BATTERY_PLUGGED_USB, BatteryManager.BATTERY_PLUGGED_WIRELESS -> info.isCharging =
                    true
            }
        }
        return info
    }

    internal class BatteryInfo {
        var batteryLevel = 0
        var isCharging = false
    }

    private fun networkType(): String? {
        var netType: String? = "Unknown"
        val connectManager =
            (applicationContext.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager)
        val telephonyManager = (applicationContext
            .getSystemService(TELEPHONY_SERVICE) as TelephonyManager)
        val networks = connectManager.allNetworks
        for (network in networks) {
            val networkCapabilities = connectManager.getNetworkCapabilities(network)!!
            if (!networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) {
                if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                    netType = "WIFI"
                    break
                }
                if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
                    if (ActivityCompat.checkSelfPermission(
                            applicationContext,
                            Manifest.permission.READ_PHONE_STATE
                        ) != PackageManager.PERMISSION_GRANTED
                    ) {
                        Log.i("get_network_type", "No permission.")
                        return netType
                    }
                    netType = checkCellularNetworkType(telephonyManager.dataNetworkType)
                }
                if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH)) {
                    netType = "Bluetooth"
                }
                if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
                    netType = "Ethernet"
                }
            }
        }
        return netType
    }
    private fun checkCellularNetworkType(type: Int): String {
        var netType = "Unknown"
        when (type) {
            TelephonyManager.NETWORK_TYPE_NR -> netType = "NR"
            TelephonyManager.NETWORK_TYPE_LTE -> {
                netType = "LTE"
                if (radio.isLTECA()) {
                    netType += "+"
                }
                if (radio.isNRConnected()) {
                    netType += " & NR"
                }
                if (radio.isNRStandby()) {
                    netType += " (NR Standby)"
                }
            }
            TelephonyManager.NETWORK_TYPE_HSPAP, TelephonyManager.NETWORK_TYPE_EVDO_0, TelephonyManager.NETWORK_TYPE_EVDO_A, TelephonyManager.NETWORK_TYPE_EVDO_B, TelephonyManager.NETWORK_TYPE_EHRPD, TelephonyManager.NETWORK_TYPE_HSDPA, TelephonyManager.NETWORK_TYPE_HSUPA, TelephonyManager.NETWORK_TYPE_HSPA, TelephonyManager.NETWORK_TYPE_TD_SCDMA, TelephonyManager.NETWORK_TYPE_UMTS -> netType =
                "3G"
            TelephonyManager.NETWORK_TYPE_GPRS, TelephonyManager.NETWORK_TYPE_EDGE, TelephonyManager.NETWORK_TYPE_CDMA, TelephonyManager.NETWORK_TYPE_1xRTT, TelephonyManager.NETWORK_TYPE_IDEN -> netType =
                "2G"
        }
        return netType
    }
}

